tests.texlua/memoize-extract.lua

```diff
@@ -0,0 +1,682 @@
+		dir_prefix = ""
+		name_prefix, code_md5sum, context_md5sum, remaining = path:match("^(.-)("..md5pat..")%-("..md5pat..")(.-).pdf$")
+	end
+
+	if not remaining then
+		-- If the pattern didn't match, return nil
+		return nil
+	end
+
+	-- check if remaining fits the scheme
+	if remaining ~= "" and not remaining:find("^%-%d+$") then
+		return nil
+	end
+
+	-- Return the extracted components
+	return dir_prefix, name_prefix, code_md5sum, context_md5sum
+end
+
+---@param prefix string
+---@return string|nil dir_prefix
+---@return string|nil name_prefix
+local function split_prefix(prefix)
+	-- try with dir_prefix and name_prefix
+	local dir_prefix, name_prefix = prefix:match("^(.*/)(.-)$")
+	if not name_prefix then
+		-- pattern did not match -> maybe the optional dir_prefix was not given
+		dir_prefix = ""
+		name_prefix = prefix:match("^(.-)$")
+	end
+
+	if not name_prefix then
+		return nil
+	end
+
+	return dir_prefix, name_prefix
+end
+
+local function parse_args(as, defaults)
+	local formats = {latex=true, plain=true, context=true}
+	local args = defaults
+
+	local i = 1
+	while i <= #as-1 do
+		local a = as[i]:match("^%-([a-zA-Z])$")
+		if not a then
+			a = as[i]:match("^%-%-([a-zA-Z]+)$")
+		end
+
+		if a == "h" then
+			print("help") -- TODO write the help output
+			exit.succ()
+		elseif a == "V" or a == "version" then
+			print(("memoize-extract.py of Memoize %s"):format(VERSION))
+			exit.succ()
+
+		elseif a == "P" or a == "pdf" then
+			args.pdf = as[i+1]
+			i = i+2
+
+		elseif a == "p" or a == "prune" then
+			args.prune = true
+
+		elseif a == "k" or a == "keep" then
+			args.keep = true
+
+		elseif a == "f" or a == "format" then
+			args.format = as[i+1]
+			if not formats[args.format] then
+				error("")
+			end
+			i = i+2
+
+		elseif a == "f" or a == "force" then
+			args.force = true
+
+		elseif a == "q" or a == "quiet" then
+			args.quiet = true
+
+		elseif a == "m" or a == "mkdir" then
+			args.mkdir = true
+
+		else
+			error("invalid token passed '"..as[i].."'")
+		end
+		i = i+1
+	end
+	args.mmz = as[#as-0]
+
+	return args
+end
+
+-------------------------
+-- temporary pathutils --
+-- only works on unix  --
+-------------------------
+local pathlib = {}
+---@param path string
+---@return string name
+---@return string remainder
+function pathlib.name(path)
+	local r, name = path:match("^(.*)/([^/]+)/?$")
+	return name or path, r
+end
+---@param path string
+---@param name string
+---@return string
+function pathlib.with_name(path, name)
+	local _, r = pathlib.name(path)
+	if r then
+		return r.."/"..name
+	end
+	return name
+end
+---@param path string
+---@return string suffix
+---@return string remainder
+function pathlib.suffix(path)
+	local r, suffix = path:match("^(.*)%.([^./]*)$")
+	return suffix or "", r or path
+end
+---@param path string
+---@param suffix string
+---@return string
+function pathlib.with_suffix(path, suffix)
+	local _, r = pathlib.suffix(path)
+	return r.."."..suffix
+end
```

Comment:
These are currently the functions in question that need to work with windows paths as well.

Well, this stronger depends on `.` being the separator of `name` and `extension`. We still need the path-sep of course, but there is no more complicated path joining or so.